### PR TITLE
Tests - Add basic utilities for module-level numerical tests

### PR DESCRIPTION
**Pull request by yzygitzh**
Adds basic utilities for module-level numerical tests, in tests/numerical_tests folder. Including:
- modules/conftest.py: common arguments and fixtures for test classes.
- modules/test_module.py: base class to run module-level numerical tests and dump results.
- modules/test_utilities.py: revised Utils class from unit tests, to respect NVTE environmental variables from users.
- utils/module_mean_and_std.py: calculate mean and std of several module test trials.
- utils/module_similarity: calculate cosine similarity of two module test stats. The stats can be either raw value, mean or std.

### CODE REVIEW

File: tests/numerical_tests/utils/module_similarity.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import argparse
+import torch
+import json
+
+def load_stats(path):
+    return torch.load(path, map_location='cpu')
+
+def compare_stats(stats_a, stats_b):
+    result = {}
+
```
[the remark this answers is not in the document]
Comment (yzygitzh) [reply tests/numerical_tests/utils/module_similarity.py:10]: The consistency is guaranteed by previous dumping stages.

File: tests/numerical_tests/utils/module_similarity.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        list_b = stats_b[key]
+
+        cosine_similarity = [
+            torch.nn.functional.cosine_similarity(
+                a.view(-1).double(),
+                b.view(-1).double(),
+                dim=0
+            ).item()
+            for a, b in zip(list_a, list_b)
+        ]
```
[the remark this answers is not in the document]
Comment (yzygitzh) [reply tests/numerical_tests/utils/module_similarity.py:22]: The consistency is guaranteed by previous dumping stages.

File: tests/numerical_tests/utils/module_similarity.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+def main(args):
+    stats_a = load_stats(args.stats_a)
+    stats_b = load_stats(args.stats_b)
+
+    comparison_result = compare_stats(stats_a, stats_b)
+
+    with open(args.output_file, 'w') as f:
+        json.dump(comparison_result, f, indent=2)
+
+if __name__ == '__main__':
```
Comment (cp5555) [inline tests/numerical_tests/utils/module_similarity.py:39]: When do you use it?

File: tests/numerical_tests/utils/module_mean_and_std.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        std_result[key] = []
+        for s in stats:
+            mean, std = s.get_mean_std()
+            mean_result[key].append(mean)
+            std_result[key].append(std)
+
+    torch.save(mean_result, args.output_mean_file)
+    torch.save(std_result, args.output_std_file)
+
+if __name__ == '__main__':
```
Comment (cp5555) [inline tests/numerical_tests/utils/module_mean_and_std.py:68]: Why do we need main?

File: tests/numerical_tests/modules/conftest.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import os
+from pathlib import Path
+
+import pytest
+import torch
+
+
+def pytest_addoption(parser):
+    parser.addoption(
```
Comment (cp5555) [inline tests/numerical_tests/modules/conftest.py:9]: Why do we need to add required option to run the tests

File: tests/numerical_tests/modules/test_module.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        if config.bf16:
+            model = Float16Module(config, model)
+        ddp_config = DistributedDataParallelConfig(use_distributed_optimizer=True)
+        model = DistributedDataParallel(config, ddp_config, model)
+        optimizer_config = OptimizerConfig(
+            optimizer='adam',
+            bf16=config.bf16,
+            use_distributed_optimizer=True,
+            lr=1e-3,
+            clip_grad=0.0
```
Comment (cp5555) [inline tests/numerical_tests/modules/test_module.py:46]: Why is there hard code for adam, lr, etc.

**Review by cp5555 — APPROVED** [review]

File: tests/numerical_tests/modules/test_module.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+    def setup_distributed(self):
+        Utils.initialize_distributed()
+
+    def setup_parallelism(self):
+        Utils.initialize_model_parallel()
+
+    def setup_random_seed(self):
+        seed = 42
+        torch.manual_seed(seed)
```
Comment (abuccts) [inline tests/numerical_tests/modules/test_module.py:29]: do you need to set [`torch.cuda.manual_seed_all`](https://docs.pytorch.org/docs/stable/generated/torch.cuda.manual_seed_all.html#torch.cuda.manual_seed_all)? and seed in python/numpy as well

File: tests/numerical_tests/modules/test_module.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        model_parallel_cuda_manual_seed(seed)
+
+    def teardown_method(self, method):
+        Utils.destroy_model_parallel()
```
Comment (abuccts) [inline tests/numerical_tests/modules/test_module.py:33]: destroy distributed as well?

File: tests/numerical_tests/utils/module_mean_and_std.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    for key, stats in stats_dict.items():
+        mean_result[key] = []
+        std_result[key] = []
+        for s in stats:
+            mean, std = s.get_mean_std()
+            mean_result[key].append(mean)
+            std_result[key].append(std)
+
+    torch.save(mean_result, args.output_mean_file)
+    torch.save(std_result, args.output_std_file)
```
Comment (abuccts) [inline tests/numerical_tests/utils/module_mean_and_std.py:66]: why not return the value directly?
